source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+    const auto& bearer_prefix = bearerPrefix();
+    if (absl::StartsWithIgnoreCase(value, bearer_prefix)) {
+      const size_t start = bearer_prefix.length();
+      return std::string(StringUtil::ltrim(value.substr(start)));
+    }
+  }
+
+  // Check for the named query string parameter.
+  const auto path = headers.Path()->value().getStringView();
+  const auto params = Http::Utility::parseQueryString(path);
+  const auto param = params.find(parameter_name);
+  if (param != params.end()) {
+    return param->second;
+  }
+
+  return EMPTY_STRING;
+}
+
+/**
+ * primary cases:
+ * 1) user is signing out
+ * 2) /_oauth redirect
+ * 3) user is authorized
+ * 4) user is unauthorized
+ */
+Http::FilterHeadersStatus OAuth2Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+
+  // The following 2 headers are guaranteed for regular requests. The asserts are helpful when
+  // writing test code to not forget these important variables in mock requests
+  const Http::HeaderEntry* host_header = headers.Host();
+  ASSERT(host_header != nullptr);
+  host_ = host_header->value().getStringView();
+
+  const Http::HeaderEntry* path_header = headers.Path();
+  ASSERT(path_header != nullptr);
+  const absl::string_view path_str = path_header->value().getStringView();
+
+  // Keep track of how many initial 302s (to the auth server) we initially distribute.
+  // The goal is to keep this number in line with the number of successful logins at the conclusion
+  // of this filter.
+  const auto* user_agent_header = headers.UserAgent();
+  const absl::string_view user_agent_value =
+      user_agent_header != nullptr ? user_agent_header->value().getStringView() : EMPTY_STRING;
+  const std::vector<std::string> v = absl::StrSplit(user_agent_value, '.');
+  user_agent_ = v[0];
+
+  sanitizeXForwardedOauthHeaders(headers);
+
+  // We should check if this is a sign out request.
+  if (path_str == config_->signoutPath()) {
+    return signOutUser(headers);
+  }
+
+  if (canSkipOAuth(headers)) {
+    // Update the path header with the query string parameters after a successful OAuth login.
+    // This is necessary if a website requests multiple resources which get redirected to the
+    // auth server. A cached login on the authentication server side will set cookies
+    // correctly but cause a race condition on future requests that have their location set
+    // to the callback path.
+    if (absl::StartsWith(path_str, config_->callbackPath())) {
+      Http::Utility::QueryParams query_parameters = Http::Utility::parseQueryString(path_str);
+
+      const auto state =
+          absl::StrReplaceAll(query_parameters.at(queryParamsState()), unescapedReplacements());
+      Http::Utility::Url state_url;
+      if (!state_url.initialize(state, false)) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      // Avoid infinite redirect storm
```

Comment:
i would maybe add a bit detail, e.g. `reject this request if we've already redirected it to prevent redirect storm`